pybop/_problem.py

```diff
@@ -44,29 +96,62 @@ def __init__(
         if len(self._target) != len(self._time_data):
             raise ValueError("Time data and signal data must be the same length.")
 
-        # Set bounds
-        self.bounds = dict(
-            lower=[param.bounds[0] for param in self.parameters],
-            upper=[param.bounds[1] for param in self.parameters],
+        # Build the model
+        if self._model._built_model is None:
+            self._model.build(
+                dataset=self._dataset,
+                fit_parameters=self.fit_parameters,
+                check_model=self.check_model,
+                init_soc=self.init_soc,
+            )
+
+    def evaluate(self, parameters):
+        """
+        Evaluate the model with the given parameters and return the signal.
+        """
+
+        y = np.asarray(self._model.simulate(inputs=parameters, t_eval=self._time_data))
+
+        return y
+
+    def evaluateS1(self, parameters):
+        """
+        Evaluate the model with the given parameters and return the signal and
+        its derivatives.
+        """
+        for i, key in enumerate(self.fit_parameters):
+            self.fit_parameters[key] = parameters[i]
+
+        y, dy = self._model.simulateS1(
+            inputs=self.fit_parameters,
+            t_eval=self._time_data,
         )
 
-        # Sample from prior for x0
-        if x0 is None:
-            self.x0 = np.zeros(self.n_parameters)
-            for i, param in enumerate(self.parameters):
-                self.x0[i] = param.rvs(1)
-        elif len(x0) != self.n_parameters:
-            raise ValueError("x0 dimensions do not match number of parameters")
+        return (np.asarray(y), np.asarray(dy))
 
-        # Add the initial values to the parameter definitions
-        for i, param in enumerate(self.parameters):
-            param.update(value=self.x0[i])
+
+class DesignProblem(BaseProblem):
+    """
+    Defines a PyBOP single output problem, follows the PINTS interface.
+    """
+
+    def __init__(
+        self,
+        parameters,
+        experiment,
+        model=None,
+        check_model=True,
+        init_soc=None,
+        x0=None,
+    ):
+        super().__init__(parameters, model, check_model, init_soc, x0)
+        self.experiment = experiment
+        self._target = None
 
         # Set the fitting parameters and build the model
-        self.fit_parameters = {o.name: o.value for o in parameters}
         if self._model._built_model is None:
```

Comment:
I think the logic for building the model should be added here instead of the BaseModel below. At this level we can add logic to trigger a build based on the value of `experiment`. This avoids modifying `build()`'s logic.